### ISSUE TITLE
FTPClient failed commands return a FirstError

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7271]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| orbitz reported|
|-|-|
|Trac ID|trac#1160|
|Type|defect|
|Created|2005-09-06 21:01:02Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1160 1160
type__defect defect
reporter__orbitz orbitz
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1126040462000000 1126040462000000
changetime__1175411415000000 1175411415000000
version__ 
owner__ 
cc__spiv cc__orbitz cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|orbitz commented|
|-|-|

```
#!html
<pre>
A failed FTPClient command returns something like:
FirstError(&lt;twisted.python.failure.Failure twisted.internet.defer.FirstError>, 0)

Is this what one wants to be returned?  Is there any good reason to return that
object rather than the error?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
No, it's not intentional.  The reason is that the system I wrote the FTP client
for didn't care about details of errors, just whether one had happened or not.

It's a bug in the API, and should be fixed.  I thought there was already a bug
on this, but I'm clearly confusing my mental todo list with the issue tracker :)

FirstError is just an unfortunate implementation details that users of FTPClient
shouldn't have to know about.

And, of course, there should be tests for the fixed behaviour added too. :)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|orbitz commented|
|-|-|

```
#!html
<pre>
Is a FirstError what a DeferredList gives when fireOnError is set?  Why 
is it a FirstError in a FirstError as well?  To fix, does the inner 
error have to be extracted or is it a simple option?

On Sep 6, 2005, at 9:05 PM, Andrew Bennetts [[Twisted issue tracker]] 
wrote:

>
> Andrew Bennetts &lt;andrew-twistedbugs@...> added the comment:
>
> No, it's not intentional.  The reason is that the system I wrote the 
> FTP client
> for didn't care about details of errors, just whether one had happened 
> or not.
>
> It's a bug in the API, and should be fixed.  I thought there was 
> already a bug
> on this, but I'm clearly confusing my mental todo list with the issue 
> tracker :)
>
> FirstError is just an unfortunate implementation details that users of 
> FTPClient
> shouldn't have to know about.
>
> And, of course, there should be tests for the fixed behaviour added 
> too. :)
>
> ----------
> status: new -> chatting
>
> _________________________________________________________
> Twisted issue tracker &lt;twisted.roundup@...>
> &lt;http://twistedmatrix.com/bugs/#1160>
> _________________________________________________________
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Right, FirstError is from a DeferredList with fireOnOneErrback set.

And there's two layers of them because inside that part of FTPClient there's two
layers of DeferredList.

Inside those two FirstErrors there's the failure that caused the original
problem.  Without looking into the code (which I don't have time for right at
this moment) I couldn't say if that's the right error to hand to the caller of
FTPClient or not.  If it is, it's just a simple matter of extracting it -- and
writing tests.

Also, I wonder if there may be some value in adding a helper somewhere to make
unwrapping FirstErrors easier, if that is what is needed here.  Perhaps an
'unwrap' method to FirstError?  Use would look like:

    dl = DeferredList(..., fireOnOneErrback=True)
    def unwrapFirstError(failure):
        failure.trap(FirstError)
        failure.unwrap()  # would raise the inner error
    dl.addErrback(unwrapFirstError)

Although even that's a lot of boilerplate... perhaps an "unwrapFirstError=True"
flag to DeferredList instead?  Or unwrapFirstError could be helper function that
lives in defer.py (or as a classmethod on FirstError?)...  Anyway, something to
think about later, after we get some experience with using FirstErrors.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Fixed in r14444, for the passive case anyway.  The fix for the non-passive case
looks trivial, but really should have a test too.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner<br>@therve set status to `new`|
|-|-|

I added a test for non-passive. In the process I fixed the problem in [#11434](https://github.com/twisted/twisted/issues/11434).

Ready to review in ftp-client-error-1160+1107.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_7"></a>|@jerub set owner to @therve|
|-|-|

Hey, this looks excellent. I really wish there wasn't so much verbage in those unit tests - you kinda lose the point of the test when reading it. The Docstring is accurate, and that's what counts.

I think this looks good, I'm okay with it being merged.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [19930]) Merge ftp-client-error-1160+1107

Author: therve
Reviewer: Jerub
Fixes [#7271](https://github.com/twisted/twisted/issues/7271)
Fixes [#11434](https://github.com/twisted/twisted/issues/11434)

Unwrap the FirtError exception from FTPClient failed commands in non-passive
mode, and fire errback of current command (usually happen when losing a
connection during a transfer). Add tests for this situation.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [19931]) Revert r19930: failure in web2 tests.

Refs [#7271](https://github.com/twisted/twisted/issues/7271)
Refs [#11434](https://github.com/twisted/twisted/issues/11434)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @jerub|
|-|-|

There was a problem with another ftp tests, that appears only when the deferred was garbage collected. I corrected the test, it looks fine.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_13"></a>|@jerub set owner to @therve|
|-|-|

Nasty, thanks for fixing that up. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set status to `closed`|
|-|-|

(In [19936]) Merge ftp-client-error-1160+1107

Author: therve
Reviewer: jerub
Fixes [#7271](https://github.com/twisted/twisted/issues/7271)
Fixes [#11434](https://github.com/twisted/twisted/issues/11434)

Unwrap the FirtError exception from FTPClient failed commands in non-passive
mode, and fire errback of current command (usually happen when losing a
connection during a transfer). Add tests for this situation, and fix a test.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
